### PR TITLE
Flexbox in WinUI draft PR

### PR DESCRIPTION
**Pull request by MatthewMachuca**
## This is a Work In Progress Draft PR
No need for an in-depth review at the moment, but comments are always appreciated.

## API Spec

https://github.com/MatthewMachuca/FlexboxApiSPEC/blob/main/flexboxAPIspec.md

## Description

Introduce Flexible Box Layout (Flexbox) to WinUI2

Achieved by creating a new subclass building off of the [Layout Class](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.layout?view=winui-2.8) currently available in WinUI2 that would enable the core functionality that follows “Flexbox Layout” guidelines

## Motivation and Context

Flexbox Layout is designed to provide an efficient way to layout, align, and distribute space among items within a container, specifically when the size of the device is either unknown or dynamic.

Flexbox is a widely used layout schematic in web development, it allows developers to quickly create complex layouts with a standardized system. By introducing Flexbox to WinUI we can increase the layout choices developers have and cater to those who desire to use this popular layout in their WinUI applications.

#### Sizing
CSS accepts many values to define sizing "px" "mm" and "%" are some examples. They can be divided into two disticnt categories. Absolute Sizing and Relative Sizing. Currently Flex Basis only accepts Numbers as a string, these values are interpreted as "px" as is the case in XAML.

In the future, adding a new property that would allow the user to define whether the Flex-Basis sizing is "px" "%" etc. This would allow for additional sizing options to be added in the future without the need to parse the data.

#### Flex Wrap

Traditionally in CSS, if an item doesn't fit and items are not instructed to "flex-shrink" the items will overflow the container.
Currently the implementation follows the rule of "if the item doesn't fit, omit" ie; if selecting NoWrap as the Wrap value, items will be omitted if they do not fit their container.

#### Flex Shrink 

Currently Flex Shrink is not implemented, items in Flexbox should have a default flex-shrink value of 1, meaning all items in a container will attempt to shrink evenly to fit.

#### Core Layout Functionality

[FlexboxLayout.cpp](https://github.com/microsoft/microsoft-ui-xaml/pull/7393/files#diff-386eb7a75e6b13097e757f697b031c9f9537bb5c421a48b704e696a40fce7085)
[FlexboxLayout.h](https://github.com/microsoft/microsoft-ui-xaml/pull/7393/files#diff-ce33984dabe55731cc93cc03a45565d0c02f4fdcffeb6ae9ef6f5846bca44583)
[FlexboxLayout.cpp](https://github.com/microsoft/microsoft-ui-xaml/pull/7393/files#diff-2cfb28b72a02edc7f1e02b0f8963c1c903f76b4af2131b7830d276976de5f38a)

#### Definitions
[ItemsReapeter.idl](https://github.com/microsoft/microsoft-ui-xaml/pull/7393/files#diff-849551b48de2fb62d79a67fafb483da720d231bf66f461f03d406958eaf7bef6)

#### Test Pages
[./Samples](https://github.com/microsoft/microsoft-ui-xaml/pull/7393/files#diff-70d41c818b61a0aacc4863ba5c38c8b78133c6b19f0e919944dead60494e8b87)

## How Has This Been Tested?

Manually tested and reviewed via Sample Pages. Compared behavior to vanilla HTML and CSS files.

| Attribute | Status |
|---|---|
| justify_content flex_start | ✔️ |
| justify_content flex_end | ✔️ |
| justify_content center | ✔️ |
| justify_content space_around | ✔️ |
| justify_content space_between | ✔️ |
| justify_content space_evenly | ✔️ |
| align_content start | ✔️ |
| align_content end | ✔️ |
| align_content center | ✔️ |
| align_content space_around | ✔️ |
| align_content space_between | ✔️ |
| align_content space_evenly | ✔️ |
| align_content stretch | ✔️ |
| align_items flex_start | ✔️ |
| align_items flex_end | ✔️ |
| align_items center | ✔️|
| align_items stretch | ✔️ |
| align_self flex_start | ✔️ |
| align_self flex_end | ✔️ |
| align_self center | ✔️ |
| align_self stretch | ✔️ |
| size relative | ❌|
| size absolute | ✔️ |
| direction column | ✔️ |
| direction column_reverse | ⚠️ |
| direction row | ✔️ |
| direction row_reverse | ⚠️ |
| wrap no_wrap | ✔️ |
| wrap wrap | ✔️ |
| wrap wrap_reverse | ⚠️ |
| grow | ✔️ |
| shrink | ❌ |
| order | ✔️|
| basis | ✔️ |

### CODE REVIEW

File: MUXControls.sln [hunk 1]
```diff
@@ -611,6 +611,12 @@ Project("{D954291E-2A0B-460D-934E-DC6B0785DB48}") = "ProgressRing_APITests", "de
 EndProject
 Project("{D954291E-2A0B-460D-934E-DC6B0785DB48}") = "ProgressBar_APITests", "dev\ProgressBar\APITests\ProgressBar_APITests.shproj", "{DD024665-31DA-4290-993E-2E46EB370C77}"
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "TestUI", "TestUI", "{EEED120B-6E82-4AF8-A995-B06CF67AFCD4}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "flextest", "flextest", "{5E02FF43-945B-40FC-B78B-7B52A6B072C3}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "WpfApp1", "WpfApp1\WpfApp1.csproj", "{5C57D7CF-307D-4400-8CB9-F1D1B25A457A}"
+EndProject
```
Comment (ranjeshj) [inline MUXControls.sln:619]: these look like projects that are not part of the PR.

File: dev/Repeater/ItemsRepeater.idl [hunk 2]
```diff
@@ -442,178 +510,219 @@ unsealed runtimeclass StackLayout : VirtualizingLayout
...
+
+     [MUX_PREVIEW]
+     [webhosthidden]
+     [MUX_PROPERTY_CHANGED_CALLBACK(TRUE)]
+     [MUX_PROPERTY_CHANGED_CALLBACK_METHODNAME("OnPropertyChanged")]
+     unsealed runtimeclass FlexboxLayout : NonVirtualizingLayout
+     {
+         FlexboxLayout();
+
+         FlexboxWrap Wrap{ get; set; };
```
Comment (ranjeshj) [inline dev/Repeater/ItemsRepeater.idl:695]: these should be dependency properties if we want bindings to work. See Spacing property of StackLayout.

File: dev/Repeater/TestUI/Samples/LayoutSamples/NonVirtualLayoutPages/NonVirtualStackLayoutSamplePage.xaml [hunk 3]
```diff
@@ -5,13 +5,232 @@
...
-        <ScrollViewer>
-            <controls:ItemsRepeater x:Name="repeater">
-                <controls:ItemsRepeater.Layout>
-                    <local:NonVirtualStackLayout />
-                </controls:ItemsRepeater.Layout>
-            </controls:ItemsRepeater>
-        </ScrollViewer>
-    </Grid>
+   <!-- <StackPanel>
+        <StackPanel.Resources>
```
Comment (ranjeshj) [inline dev/Repeater/TestUI/Samples/LayoutSamples/NonVirtualLayoutPages/NonVirtualStackLayoutSamplePage.xaml:9]: you should create a separate page for flex layout since this file is for NonVirtualStackLayout

File: test/MUXControlsTestApp/MainPage.xaml [hunk 4]
```diff
@@ -91,6 +91,10 @@
                     </ItemsControl.ItemTemplate>
                 </ItemsControl>
             </ScrollViewer>
+            <Button> Test Button</Button>
```
Comment (ranjeshj) [inline test/MUXControlsTestApp/MainPage.xaml:94]: looks unrelated. please revert

File: dev/Repeater/TestUI/Samples/FlexboxAlignContentPage.xaml [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+    xmlns:local="using:MUXControlsTestApp.Samples"
+    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
+    xmlns:controls="using:Microsoft.UI.Xaml.Controls"
+    mc:Ignorable="d"
+    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">
+
+    <StackPanel>
```
Comment (ranjeshj) [inline dev/Repeater/TestUI/Samples/FlexboxAlignContentPage.xaml:12]: Consider adding API tests - see flowlayout tests for examples  on how to write layout tests. [https://github.com/microsoft/microsoft-ui-xaml/blob/main/dev/Repeater/APITests/FlowLayoutTests.cs](https://github.com/microsoft/microsoft-ui-xaml/blob/main/dev/Repeater/APITests/FlowLayoutTests.cs)

File: dev/Repeater/FlexboxLayout.cpp [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+        // TODO: What about flex-shrink? Should we try them with less?
+        // TODO: This is where flex-basis would come into play
+        child.Measure(availableSize);
+        winrt::Size childDesiredSize = child.DesiredSize();
+
+      //flexbasis
+    //Asign initial size based on FlexBasis
+        float flexBasis = 0;
+
+        float inputBasis = float((GetBasis(child)));
```
Comment (chiaramooney) [inline dev/Repeater/FlexboxLayout.cpp:208]: don't need the extra inputBasis variable here. if GetBasis returns a negative float, no logic will run so we don't need the extra value. 

also a general C++ rule "almost always use auto" i.e. we can change line 208 to `auto flexBasis = float((GetBasis(child)));`

File: dev/Repeater/FlexboxLayout.cpp [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+
+        
+        if (inputBasis > 0) {
+            flexBasis = float((GetBasis(child)));
+            childDesiredSize = CreateSize(inputBasis, CrossAxis(childDesiredSize));
```
Comment (chiaramooney) [inline dev/Repeater/FlexboxLayout.cpp:213]: note our logic here only handles the cases where flexDirection is 'row'. We will want to extend this logic to also handle the case where flexDirection is column and the flexBasis value will apply to the items height.

File: dev/Repeater/FlexboxLayout.cpp [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+        // Give each child the maximum available space
+        // TODO: What about flex-shrink? Should we try them with less?
+        // TODO: This is where flex-basis would come into play
+        child.Measure(availableSize);
+        winrt::Size childDesiredSize = child.DesiredSize();
+
+      //flexbasis
+    //Asign initial size based on FlexBasis
+
+        auto flexBasis = float((GetBasis(child)));
```
Comment (chrisglein) [inline dev/Repeater/FlexboxLayout.cpp:208]: Why are you needing a type cast here (to float)?
Also, if you're casting, it should be a static_cast most likely.

File: dev/Repeater/FlexboxLayout.cpp [hunk 9]
```diff
@@ -0,0 +1,489 @@
...
+        case winrt::FlexboxAlignItems::End:
+            crossOffset += excessCrossAxisInRow;
+            break;
+        case winrt::FlexboxAlignItems::Stretch:
+            childDesiredSize = CreateSize(MainAxis(childDesiredSize), (usedInCurrentCrossAxis > 0 ? usedInCurrentCrossAxis : info.CrossAxis));
+            break;
+        }
+
+        // In Reversed mode we need to swap the coordinates so that items grom from right/bottom to left/top
+        // TODO: Using finalSize here means things will become right/bottom aligned, which doesn't seem right
```
Comment (chrisglein) [inline dev/Repeater/FlexboxLayout.cpp:469]: Do we have a test to answer whether this behavior is correct or not?

File: dev/Repeater/FlexboxLayout.h [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+    winrt::FlexboxAlignItems AlignItems();
+    void AlignItems(winrt::FlexboxAlignItems const& value);
+
+    winrt::FlexboxAlignContent AlignContent();
+    void AlignContent(winrt::FlexboxAlignContent const& value);
+
+    //flexbasis
+   // winrt::FlexboxBasis FlexBasis();
+    //void FlexBasis(winrt::FlexboxBasis const& value);
```
Comment (chrisglein) [inline dev/Repeater/FlexboxLayout.h:40]: What's the status on this commented out property?

File: dev/Repeater/FlexboxLayout.h [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    bool IsReversed();
+    bool IsWrapping();
+    float MainAxis(winrt::Size const& value);
+    float CrossAxis(winrt::Size const& value);
+   // int FlexBasis(winrt::Size const& value);
+    winrt::Size CreateSize(float mainAxis, float crossAxis);
+    winrt::Point CreatePoint(float mainAxis, float crossAxis);
+    std::vector<winrt::UIElement> ChildrenSortedByOrder(winrt::NonVirtualizingLayoutContext const& context);
+    // testing for basis initial
+    std::vector<winrt::UIElement> ChildrenInitial(winrt::NonVirtualizingLayoutContext const& ci);
```
Comment (chrisglein) [inline dev/Repeater/FlexboxLayout.h:67]: Is this temporary code or something that will survive to the final implementation?

File: dev/Repeater/ItemsRepeater.idl [hunk 12]
```diff
@@ -442,178 +510,219 @@ unsealed runtimeclass StackLayout : VirtualizingLayout
...
-   // overridable void OnElementMeasured(Windows.UI.Xaml.UIElement element, Int32 index, Windows.Foundation.Size availableSize, Windows.Foundation.Size measureSize, Windows.Foundation.Size desiredSize, Windows.Foundation.Size provisionalArrangeSize, VirtualizingLayoutContext context);
+        // Removing until we are ready to expose.
+        // overridable FlowLayoutAnchorInfo GetAnchorForRealizationRect(Windows.Foundation.Size availableSize, VirtualizingLayoutContext context);
+        // overridable Windows.Foundation.Rect GetExtent(Windows.Foundation.Size availableSize, VirtualizingLayoutContext context, Windows.UI.Xaml.UIElement firstRealized, Int32 firstRealizedItemIndex, Windows.Foundation.Rect firstRealizedLayoutBounds, Windows.UI.Xaml.UIElement lastRealized, Int32 lastRealizedItemIndex, Windows.Foundation.Rect lastRealizedLayoutBounds);
+        // overridable void OnElementMeasured(Windows.UI.Xaml.UIElement element, Int32 index, Windows.Foundation.Size availableSize, Windows.Foundation.Size measureSize, Windows.Foundation.Size desiredSize, Windows.Foundation.Size provisionalArrangeSize, VirtualizingLayoutContext context);
```
Comment (chrisglein) [inline dev/Repeater/ItemsRepeater.idl:516]: Unintended change in whitespace?

Comment (chrisglein) [reply dev/Repeater/ItemsRepeater.idl:516]: Something is off here in the diff and causing it to look like more code changed than actually did (see parts that follow). Can you figure that out?

File: dev/Repeater/ItemsRepeater.idl [hunk 13]
```diff
@@ -442,178 +510,219 @@ unsealed runtimeclass StackLayout : VirtualizingLayout
...
+         void ClearSelection();
+
+         protected void OnPropertyChanged(String propertyName);
+     }
+
+     [MUX_PREVIEW]
+     [webhosthidden]
+     [MUX_PROPERTY_CHANGED_CALLBACK(TRUE)]
+     [MUX_PROPERTY_CHANGED_CALLBACK_METHODNAME("OnPropertyChanged")]
+     unsealed runtimeclass FlexboxLayout : NonVirtualizingLayout
```
Comment (michael-hawker) [inline dev/Repeater/ItemsRepeater.idl:691]: I'm not too familiar with idl, can these be put in a separate file so this isn't as massive and in the future it becomes clearer which things are changing specific layout implementation contracts vs. the base ItemsRepeater one?

Comment (chingucoding) [reply dev/Repeater/ItemsRepeater.idl:691]: I totally agree with @michael-hawker , I think it would be better to split off a new (or several new) IDL files for layouts.

File: dev/Repeater/FlexboxLayout.cpp (hunk 8 above)
Comment (MatthewMachuca) [reply dev/Repeater/FlexboxLayout.cpp]: The compiler have me errors trying to += it without casting it to a float.

It's been changed to static cast now.

File: dev/Repeater/FlexboxLayout.h (hunk 10 above)
Comment (MatthewMachuca) [reply dev/Repeater/FlexboxLayout.h]: removed.

File: dev/Repeater/FlexboxLayout.h [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    static void OnChildPropertyChanged(
+        const winrt::DependencyObject& sender,
+        const winrt::DependencyPropertyChangedEventArgs& args);
+
+    void OnPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args);
+
+private:
+    bool IsHorizontal();
+    bool IsReversed();
+    bool IsWrapping();
```
Comment (asklar) [inline dev/Repeater/FlexboxLayout.h:55]: these methods can be const

File: dev/Repeater/FlexboxLayout.h [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    float MainAxis(winrt::Size const& value);
+    float CrossAxis(winrt::Size const& value);
+    winrt::Size CreateSize(float mainAxis, float crossAxis);
+    winrt::Point CreatePoint(float mainAxis, float crossAxis);
```
Comment (asklar) [inline dev/Repeater/FlexboxLayout.h:59]: these too can be const

File: dev/Repeater/FlexboxLayoutState.h [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+#pragma once
+
+class FlexboxLayoutState :
+    public winrt::implements<FlexboxLayoutState, winrt::IInspectable>
+{
+public:
+    struct RowMeasureInfo
+    {
+    public:
```
Comment (asklar) [inline dev/Repeater/FlexboxLayoutState.h:12]: remove `public`, structs are public by default

File: dev/Repeater/FlexboxLayoutState.h [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+        float MainAxis;
+        float CrossAxis;
+        int Count;
+        float Grow;
+        float Shrink;
+        float Basis;
```
Comment (asklar) [inline dev/Repeater/FlexboxLayoutState.h:18]: initialize these as values are not zero-initialized in C++.
Name them according to the guidelines (m_ for members)
```suggestion
        float m_mainAxis{};
        float m_crossAxis{};
        int m_count{};
        float m_grow{};
        float m_shrink{};
        float m_basis{};
```

File: dev/Repeater/ItemsRepeater.idl (hunk 13 above)
Comment (asklar) [inline dev/Repeater/ItemsRepeater.idl:691]: why does this need to be unsealed?

File: dev/Repeater/TestUI/RepeaterTestUIPage.xaml.cs [hunk 18]
```diff
@@ -122,6 +122,46 @@ public RepeaterTestUIPage()
                 Frame.NavigateWithoutAnimation(typeof(MUXControlsTestApp.Samples.ScaleAnimatedVerticalListDemo));
             };
 
+            flexboxDemo.Click += delegate {
+                Frame.NavigateWithoutAnimation(typeof(FlexboxAlignItemsPage));
+            };
```
Comment (asklar) [inline dev/Repeater/TestUI/RepeaterTestUIPage.xaml.cs:127]: JFYI that you can write these more concisely as:
```cs
            flexboxDemo.Click += () => Frame.NavigateWithoutAnimation(typeof(FlexboxAlignItemsPage));
```

File: dev/Repeater/TestUI/Samples/FlexboxWrapPage.xaml [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+            </Grid>
+            <Grid MinWidth="60">
+                <TextBlock>7</TextBlock>
+            </Grid>
+            <Grid MinWidth="100">
+                <TextBlock>8</TextBlock>
+            </Grid>
+        </controls:LayoutPanel>
+
+            <controls:WebView2 Source="https://codepen.io/matthewmachuca/pen/poLLeGM" Style="{StaticResource }" ></controls:WebView2>
```
Comment (asklar) [inline dev/Repeater/TestUI/Samples/FlexboxWrapPage.xaml:76]: is this webview showing anything? if so, do we need the style tag?

File: dev/Repeater/FlexboxLayout.cpp [hunk 20]
```diff
@@ -0,0 +1,468 @@
...
+        });
+
+    return sorted;
+}
+
+void FlexboxLayout::InitializeForContextCore(winrt::LayoutContext const& context)
+{
+    auto state = context.LayoutState();
+    winrt::com_ptr<FlexboxLayoutState> flexboxState = nullptr;
+    if (state)
```
Comment (asklar) [inline dev/Repeater/FlexboxLayout.cpp:126]: why does this need to be a com_ptr, can you just use ```cpp
FlexboxLayoutState flexboxState{nullptr};
```

Comment (asklar) [reply dev/Repeater/FlexboxLayout.cpp:126]: actually can just replace the whole thing with:
```cpp
auto flexboxState = state.try_as<FlexboxLayoutState>();
```

File: dev/Repeater/FlexboxLayout.cpp [hunk 21]
```diff
@@ -0,0 +1,468 @@
...
+    };
+   
+    
+
+    std::vector<winrt::UIElement> sortedChildren = ChildrenSortedByOrder(context.try_as<winrt::NonVirtualizingLayoutContext>());
+    for (winrt::UIElement const& child : sortedChildren)
+    {
+        
+
+        // Give each child the maximum available space
```
Comment (asklar) [inline dev/Repeater/FlexboxLayout.cpp:195]: remove extra whitespace lines here and others

File: dev/Repeater/FlexboxLayout.cpp [hunk 22]
```diff
@@ -0,0 +1,468 @@
...
+winrt::Size FlexboxLayout::MeasureOverride(
+    winrt::NonVirtualizingLayoutContext const& context,
+    winrt::Size const& availableSize)
+{
+    auto state = context.LayoutState().as<FlexboxLayoutState>();
+    state->Rows.clear();
+
+    unsigned int itemsInRow = 0;
+    float growInRow = 0.0;
+    float shrinkInRow = 0.0;
```
Comment (asklar) [inline dev/Repeater/FlexboxLayout.cpp:157]: this and others are float, but the properties they get set from are double. Use double where the value is double :)

File: dev/Repeater/FlexboxLayout.cpp [hunk 23]
```diff
@@ -0,0 +1,468 @@
...
+                }
+            }
+     
+       
+        // Contribute our space
+        usedInCurrentMainAxis += MainAxis(childDesiredSize);
+        usedInCurrentCrossAxis = std::max(usedInCurrentCrossAxis, CrossAxis(childDesiredSize));
+        itemsInRow++;
+        growInRow += float(GetGrow(child));
+
```
Comment (asklar) [inline dev/Repeater/FlexboxLayout.cpp:234]: cast goes away when you make these doubles

File: dev/Repeater/FlexboxLayout.cpp [hunk 24]
```diff
@@ -0,0 +1,468 @@
...
+}
+
+void FlexboxLayout::OnPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+}
+
+void FlexboxLayout::OnChildPropertyChanged(const winrt::DependencyObject& sender, const winrt::DependencyPropertyChangedEventArgs& args)
+{
+}
+
```
Comment (asklar) [inline dev/Repeater/FlexboxLayout.cpp:468]: sounds like these should call Invalidate ?

File: dev/Repeater/TestUI/Samples/FlexboxWrapPage.xaml (hunk 19 above)
Comment (MatthewMachuca) [reply dev/Repeater/TestUI/Samples/FlexboxWrapPage.xaml:76]: The webview displays a visual test comparison. it links to a codepenn of html/css example layout. it's currently my solution to having a side by side test.

the style makes the webview large enough to view

File: dev/Repeater/FlexboxLayout.cpp (hunk 22 above)
Comment (MatthewMachuca) [reply dev/Repeater/FlexboxLayout.cpp:157]: Oops, guess that's why I had to cast my variables in another spot

File: dev/Repeater/FlexboxLayout.cpp (hunk 23 above)
Comment (MatthewMachuca) [reply dev/Repeater/FlexboxLayout.cpp:234]: I keep getting a compiler error that states that "potential loss of data converting double to float" 

This is why I originally casted the value as a float before any computations.

File: dev/Repeater/ItemsRepeater.idl (hunk 13 above)
Comment (chiaramooney) [reply dev/Repeater/ItemsRepeater.idl:691]: cc @chrisglein you'll probably have the context here instead of Matthew

File: dev/Repeater/TestUI/Samples/FlexboxOrderPage.xaml [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+        </controls:LayoutPanel>
+
+        <controls:LayoutPanel>
+            <controls:LayoutPanel.Layout>
+                <controls:FlexboxLayout/>
+            </controls:LayoutPanel.Layout>
+            <Grid controls:FlexboxLayout.Order="5">
+                <TextBlock>5</TextBlock>
+            </Grid>
+            <Grid controls:FlexboxLayout.Order="-1">
```
Comment (chiaramooney) [inline dev/Repeater/TestUI/Samples/FlexboxOrderPage.xaml:64]: Just to confirm have we validated against CSS what the negative case is supposed to be? Should it be placed first or are negative numbers ignored and assumed to be 0?

File: dev/Repeater/TestUI/Samples/FlexboxAlignContentPage.xaml [hunk 26]
```diff
@@ -0,0 +1,135 @@
+﻿<Page
```
Comment (chiaramooney) [inline dev/Repeater/TestUI/Samples/FlexboxAlignContentPage.xaml:1]: I'm not seeing a test page for flex basis. We'll want to add a page specifically for that property even if it is being used in other pages.

File: dev/Repeater/ItemsRepeater.idl (hunk 13 above)
Comment (MikeHillberg) [reply dev/Repeater/ItemsRepeater.idl:691]: Default should be unsealed unless there's special reason to seal it. (Developers hate sealed classes) We have a bunch of sealed classes in Xaml because of a technical limitation in early Silverlight, but there's a backlog item to unseal them.

Comment (asklar) [reply dev/Repeater/ItemsRepeater.idl:691]: unsealed causes build times to go through the roof for cppwinrt due to the heavy crtp involved, no?

Comment (MikeHillberg) [reply dev/Repeater/ItemsRepeater.idl:691]: Even if you're just using it and not subclassing? Most of Xaml is unsealed today.

Comment (asklar) [reply dev/Repeater/ItemsRepeater.idl:691]: I _think_ that's true (and why compiling xaml code is slow), but @Scottj1s can keep me honest

File: dev/Repeater/TestUI/Samples/FlexboxGrowPage.xaml [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+            <Style TargetType="TextBlock">
+                <Setter Property="Foreground" Value="White"/>
+                <Setter Property="VerticalAlignment" Value="Center"/>
+                <Setter Property="HorizontalAlignment" Value="Center"/>
+                <Setter Property="FontWeight" Value="Bold"/>
+                <Setter Property="FontSize" Value="24"/>
+            </Style>
+            <Style TargetType="Grid">
+                <Setter Property="Background" Value="Orange"/>
+             <!--   <Setter Property="MinWidth" Value="80"/> -->
```
Comment (chrisglein) [inline dev/Repeater/TestUI/Samples/FlexboxGrowPage.xaml:23]: I see that this line was commented out from the previous version of these test files. And replaced with `Basis`. What was the behavior you were seeing with `MinWidth` that caused you to remove it? I think there may be something interesting to this, as `MinWidth` is a XAML intrinsic layout value, and this likely fits into the "How do XAML layout and Flexbox layout combine when there are overlapping concepts?"

File: dev/Repeater/FlexboxLayout.cpp [hunk 28]
```diff
@@ -0,0 +1,460 @@
...
+void FlexboxLayout::InitializeForContextCore(winrt::LayoutContext const& context)
+{
+    auto state = context.LayoutState();
+    auto flexboxState = state.try_as<FlexboxLayoutState>();
+    if (state)
+    {
+        flexboxState = state.as<FlexboxLayoutState>();
+    }
```
Comment (chrisglein) [inline dev/Repeater/FlexboxLayout.cpp:129]: Now that you've switched to `try_as`, you should be able to delete these lines.

File: dev/Repeater/FlexboxLayout.cpp [hunk 29]
```diff
@@ -0,0 +1,460 @@
...
+winrt::Size FlexboxLayout::MeasureOverride(
+    winrt::NonVirtualizingLayoutContext const& context,
+    winrt::Size const& availableSize)
+{
+    auto state = context.LayoutState().as<FlexboxLayoutState>();
+    state->Rows.clear();
+
+    unsigned int itemsInRow = 0;
+    float growInRow = 0.0;
+    float shrinkInRow = 0.0;
```
Comment (chrisglein) [inline dev/Repeater/FlexboxLayout.cpp:157]: Shrink isn't yet implemented, correct? So `shrinkInRow` isn't used?

**Review by chrisglein — CHANGES_REQUESTED** [review]

**Comment by riverar** [comment]
Microsoft today announced no new features were landing in WinUI 2.x after 2.8, so what vehicle would be used here?

**Comment by chrisglein** [comment]
> Microsoft today announced no new features were landing in WinUI 2.x after 2.8, so what vehicle would be used here? And would this port over to WinUI 3.x?

The expectation is this would target WinUI 3.x, not WinUI 2.x. Context: This was an intern project and in the interest of of their available time the intern didn't tackle the WinUI3 conversion. It's not expected that this code would merge as is; this was a mechanism for quickly making progress on the design.

**Comment by riverar** [comment]
Makes sense to me, @chrisglein!

**Review by bpulliam — APPROVED** [review]
